src/main/java/slimeknights/mantle/recipe/data/BaseRecipeProvider.java

```diff
@@ -0,0 +1,344 @@
+package slimeknights.mantle.recipe.data;
+
+import net.minecraft.advancements.ICriterionInstance;
+import net.minecraft.data.CookingRecipeBuilder;
+import net.minecraft.data.DataGenerator;
+import net.minecraft.data.IFinishedRecipe;
+import net.minecraft.data.RecipeProvider;
+import net.minecraft.data.ShapedRecipeBuilder;
+import net.minecraft.data.ShapelessRecipeBuilder;
+import net.minecraft.data.SingleItemRecipeBuilder;
+import net.minecraft.item.Item;
+import net.minecraft.item.crafting.IRecipeSerializer;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.tags.ITag;
+import net.minecraft.tags.ItemTags;
+import net.minecraft.util.IItemProvider;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.common.crafting.conditions.ICondition;
+import net.minecraftforge.common.crafting.conditions.IConditionBuilder;
+import net.minecraftforge.common.crafting.conditions.NotCondition;
+import net.minecraftforge.common.crafting.conditions.TagEmptyCondition;
+import net.minecraftforge.registries.IForgeRegistryEntry;
+import slimeknights.mantle.registration.object.BuildingBlockObject;
+import slimeknights.mantle.registration.object.MetalItemObject;
+import slimeknights.mantle.registration.object.WallBuildingBlockObject;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ * Shared logic for each module's recipe provider
+ */
+public abstract class BaseRecipeProvider extends RecipeProvider implements IConditionBuilder {
+  public static String modID;
+
+  public BaseRecipeProvider(DataGenerator generator, String modID) {
+    super(generator);
+    this.modID = modID;
+  }
+
+  @Override
+  protected abstract void registerRecipes(Consumer<IFinishedRecipe> consumer);
+
+  @Override
+  public abstract String getName();
+
+
+  /* Location helpers */
+
+  /**
+   * Gets a resource location for Tinkers
+   * @param id  Location path
+   * @return  Location for Tinkers
+   */
+  protected static ResourceLocation location(String id) {
+    return new ResourceLocation(modID, id);
+  }
+
+  /**
+   * Gets a resource location string for Tinkers
+   * @param id  Location path
+   * @return  Location for Tinkers
+   */
+  protected static String locationString(String id) {
+    return modID + ":" + id;
+  }
+
+  /**
+   * Prefixes the resource location path with the given value
+   * @param item    Item registry name to use
+   * @param prefix  Prefix value
+   * @return  Resource location path
+   */
+  protected static ResourceLocation wrap(IItemProvider item, String prefix, String suffix) {
```

Comment:
Might be worth reworking these a little. `wrap` and `wrapR` are a bit awkward. If nothing else, have `wrapI` and `wrapR`, but `wrapItem` and `wrap`/`wrapSupplier` might be better.